lib/spack/spack/installer.py

```diff
@@ -957,52 +962,82 @@ def _ensure_locked(self, lock_type, pkg):
                 otherwise ``False``
+            install_deps (bool): ``True`` if installing package dependencies,
+                otherwise ``False``
         """
         tty.debug('Initializing the build queue for {0}'.format(self.pkg.name))
         install_compilers = spack.config.get(
             'config:install_missing_compilers', False)
 
-        if install_deps:
-            for dep in self.spec.traverse(order='post', root=False):
-                dep_pkg = dep.package
-
-                # First push any missing compilers (if requested)
-                if install_compilers:
-                    self._add_bootstrap_compilers(dep_pkg)
-
-                if package_id(dep_pkg) not in self.build_tasks:
-                    self._push_task(dep_pkg, False, 0, 0, STATUS_ADDED)
+        def not_installed(spec):
+            if spec.package.installed:
+                self._update_installed(spec.package)
+                tty.debug('Assuming dependencies of {0} are also installed'
+                          .format(spec.package.name))
+                for dep in spec.dependencies():
+                    self._update_installed(dep.package)
+            return not spec.package.installed
 
-                # Clear any persistent failure markings _unless_ they are
-                # associated with another process in this parallel build
-                # of the spec.
-                spack.store.db.clear_failure(dep, force=False)
+        if install_package:
+            self._queue_pkg(self.pkg, install_deps and install_compilers)
 
-            # Push any missing compilers (if requested) as part of the
-            # package dependencies.
-            if install_compilers:
-                self._add_bootstrap_compilers(self.pkg)
-
-        if install_package and self.pkg_id not in self.build_tasks:
-            # Be sure to clear any previous failure
-            spack.store.db.clear_failure(self.pkg.spec, force=True)
+            # Don't proceed if the explicit package has already been installed
+            if self.pkg_id in self.installed:
+                tty.info('{0} is already installed'.format(self.pkg_id))
+                return
 
             # If not installing dependencies, then determine their
-            # installation status before proceeding
+            # installation status before proceeding.
             if not install_deps:
-                self._check_deps_status()
+                self._check_deps_status(install_compilers)
+                return
 
-            # Now add the package itself, if appropriate
-            self._push_task(self.pkg, False, 0, 0, STATUS_ADDED)
+        if install_deps:
+            # Traverse dependencies from the bottom-up so any that are flagged
+            # as installed can be readily removed
+            for dep in self.pkg.spec.traverse(root=False,
+                                              predicate=not_installed):
```

Comment:
I'm assuming that adding this predicate would be sufficient for avoiding the installation of dependencies when a dependent is installed (with the caveat mentioned elsewhere that generally we may still have to install non-build deps). What is the purpose of additionally marking the dependencies as installed (i.e. in the above `not_installed` function)?

Reply:
IIRC the core issue spawning this work was the undesirable affect of having dependencies of installed packages being installed.  

So flagging dependencies of installed packages as being installed should help ensure there is no attempt to install the dependencies.